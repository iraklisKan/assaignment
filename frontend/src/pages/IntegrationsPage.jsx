import React, { useState, useEffect, useCallback } from 'react';
import { integrationsAPI } from '../services/api';
import IntegrationForm from '../components/IntegrationForm.jsx';
import { 
  Button, 
  Card, 
  Badge, 
  Alert, 
  LoadingSpinner,
  Table,
  EmptyState
} from '../components/ui/index.jsx';

function IntegrationsPage() {
  const [integrations, setIntegrations] = useState([]);
  const [providers, setProviders] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [showForm, setShowForm] = useState(false);
  const [editingIntegration, setEditingIntegration] = useState(null);
  const [activeFilter, setActiveFilter] = useState('all');

  const fetchIntegrations = useCallback(async () => {
    try {
      setLoading(true);
      const params = activeFilter === 'all' ? {} : { active: activeFilter === 'active' };
      const response = await integrationsAPI.getAll(params);
      setIntegrations(response.data.data);
      setError(null);
    } catch (err) {
      setError(err.message || 'Failed to fetch integrations');
    } finally {
      setLoading(false);
    }
  }, [activeFilter]);

  const fetchProviders = useCallback(async () => {
    try {
      const response = await integrationsAPI.getProviders();
      setProviders(response.data.data);
    } catch (err) {
      console.error('Failed to load providers:', err);
    }
  }, []);

  useEffect(() => {
    fetchIntegrations();
    fetchProviders();
  }, [fetchIntegrations, fetchProviders]);

  const handleCreate = () => {
    setEditingIntegration(null);
    setShowForm(true);
  };

  const handleEdit = (integration) => {
    setEditingIntegration(integration);
    setShowForm(true);
  };

  const handleToggleActive = async (integration) => {
    const action = integration.active ? 'deactivate' : 'activate';
    if (!window.confirm(`Are you sure you want to ${action} "${integration.name}"?`)) return;

    try {
      await integrationsAPI.update(integration.id, { active: !integration.active });
      fetchIntegrations();
    } catch (err) {
      alert(`Failed to ${action} integration: ${err.message}`);
    }
  };

  const handleDelete = async (id, name) => {
    if (!window.confirm(`Are you sure you want to PERMANENTLY DELETE "${name}"? This action cannot be undone.`)) return;

    try {
      await integrationsAPI.hardDelete(id);
      fetchIntegrations();
    } catch (err) {
      alert(`Failed to delete integration: ${err.message}`);
    }
  };

  const handleFormSubmit = async (data) => {
    if (editingIntegration) {
      await integrationsAPI.update(editingIntegration.id, data);
    } else {
      await integrationsAPI.create(data);
    }
    setShowForm(false);
    setEditingIntegration(null);
    fetchIntegrations();
  };

  const handleFormCancel = () => {
    setShowForm(false);
    setEditingIntegration(null);
  };

  if (loading) {
    return (
      <div className="flex flex-col items-center justify-center py-16">
        <LoadingSpinner size="lg" />
        <p className="mt-4 text-gray-600">Loading integrations...</p>
      </div>
    );
  }

  return (
    <div className="space-y-6">
      <div className="flex items-center justify-between">
        <div>
          <h1 className="text-2xl font-bold text-gray-900">Integration Management</h1>
          <p className="mt-1 text-sm text-gray-600">
            Configure and manage your exchange rate API providers
          </p>
        </div>
        <Button onClick={handleCreate}>
          <svg className="h-5 w-5 mr-2" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M12 6v6m0 0v6m0-6h6m-6 0H6" />
          </svg>
          Add Integration
        </Button>
      </div>

      {error && (
        <Alert variant="error" onClose={() => setError(null)}>
          {error}
        </Alert>
      )}

      <Card>
        <Card.Header>
          <div className="flex items-center justify-between">
            <h2 className="text-lg font-semibold text-gray-900">Active Integrations</h2>
            <div className="flex gap-2">
              {filterOptions.map(option => (
                <button
                  key={option.value}
                  onClick={() => setActiveFilter(option.value)}
                  className={`px-3 py-1.5 text-sm font-medium rounded-lg transition-colors ${
                    activeFilter === option.value
                      ? 'bg-blue-100 text-blue-700'
                      : 'text-gray-600 hover:bg-gray-100'
                  }`}
                >
                  {option.label}
                </button>
              ))}
            </div>
          </div>
        </Card.Header>

        <Card.Content className="p-0">
          {integrations.length === 0 ? (
            <EmptyState
              icon={
                <svg className="h-12 w-12 text-gray-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                  <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M11 4a2 2 0 114 0v1a1 1 0 001 1h3a1 1 0 011 1v3a1 1 0 01-1 1h-1a2 2 0 100 4h1a1 1 0 011 1v3a1 1 0 01-1 1h-3a1 1 0 01-1-1v-1a2 2 0 10-4 0v1a1 1 0 01-1 1H7a1 1 0 01-1-1v-3a1 1 0 00-1-1H4a2 2 0 110-4h1a1 1 0 001-1V7a1 1 0 011-1h3a1 1 0 001-1V4z" />
                </svg>
              }
              title="No integrations found"
              description="Get started by adding your first exchange rate API integration"
              action={
                <Button onClick={handleCreate}>
                  Add Your First Integration
                </Button>
              }
            />
          ) : (
            <Table 
              headers={['Name', 'Provider', 'Priority', 'Poll Interval', 'Status', 'Actions']}
            >
              {integrations.map((integration) => (
                <tr key={integration.id} className="hover:bg-gray-50 transition-colors">
                  <td className="px-6 py-4 whitespace-nowrap">
                    <div className="font-medium text-gray-900">{integration.name}</div>
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-600">
                    {integration.provider}
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap">
                    <Badge variant="default">{integration.priority}</Badge>
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-600">
                    {integration.poll_interval_seconds}s
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap">
                    <Badge variant={integration.active ? 'success' : 'danger'}>
                      {integration.active ? 'Active' : 'Inactive'}
                    </Badge>
                  </td>
                  <td className="px-6 py-4 whitespace-nowrap text-sm space-x-2">
                    <Button 
                      variant="ghost" 
                      size="sm" 
                      onClick={() => handleEdit(integration)}
                    >
                      Edit
                    </Button>
                    <Button 
                      variant={integration.active ? 'warning' : 'success'}
                      size="sm" 
                      onClick={() => handleToggleActive(integration)}
                    >
                      {integration.active ? 'Deactivate' : 'Activate'}
                    </Button>
                    <Button 
                      variant="danger" 
                      size="sm" 
                      onClick={() => handleDelete(integration.id, integration.name)}
                    >
                      Delete
                    </Button>
                  </td>
                </tr>
              ))}
            </Table>
          )}
        </Card.Content>
      </Card>

      {showForm && (
        <IntegrationForm
          integration={editingIntegration}
          providers={providers}
          onSubmit={handleFormSubmit}
          onCancel={handleFormCancel}
        />
      )}
    </div>
  );
}

const filterOptions = [
  { label: 'All', value: 'all' },
  { label: 'Active', value: 'active' },
  { label: 'Inactive', value: 'inactive' },
];

export default IntegrationsPage;
