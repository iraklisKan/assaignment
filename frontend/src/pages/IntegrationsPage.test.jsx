import React from "react";
import { render } from "@testing-library/react";
import IntegrationsPage from "./IntegrationsPage.jsx";

// Mock the API service
jest.mock("../services/api", () => ({
  integrationsAPI: {
    getIntegrations: jest.fn(() => Promise.resolve({ data: [] })),
    getProviders: jest.fn(() => Promise.resolve({ data: [] })),
    createIntegration: jest.fn(() => Promise.resolve({ data: {} })),
    updateIntegration: jest.fn(() => Promise.resolve({ data: {} })),
    deleteIntegration: jest.fn(() => Promise.resolve({ data: {} })),
    testIntegration: jest.fn(() => Promise.resolve({ data: {} })),
  },
}));

describe("IntegrationsPage", () => {
  it("should render without crashing", () => {
    const { container } = render(<IntegrationsPage />);
    expect(container.firstChild).toBeInTheDocument();
  });

  // @REVIEW: you can also match snapshots, this automatically generates a snapshot file inside the __snapshots__ folder, see https://jestjs.io/docs/snapshot-testing
  it("should match snapshot", () => {
    const { container } = render(<IntegrationsPage />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
