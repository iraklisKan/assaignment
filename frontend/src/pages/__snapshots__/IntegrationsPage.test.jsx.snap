// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntegrationsPage should match snapshot 1`] = `
<div
  class="space-y-6"
>
  <div
    class="flex items-center justify-between"
  >
    <div>
      <h1
        class="text-2xl font-bold text-gray-900"
      >
        Integration Management
      </h1>
      <p
        class="mt-1 text-sm text-gray-600"
      >
        Configure and manage your exchange rate API providers
      </p>
    </div>
    <button
      class="inline-flex items-center justify-center font-medium rounded-lg transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 px-4 py-2 text-sm "
      type="button"
    >
      <svg
        class="h-5 w-5 mr-2"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
      Add Integration
    </button>
  </div>
  <div
    class="px-4 py-3 rounded-lg border bg-red-50 border-red-200 text-red-800 "
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex-1"
      >
        _api.integrationsAPI.getAll is not a function
      </div>
      <button
        class="ml-3 inline-flex text-gray-400 hover:text-gray-600 focus:outline-none"
      >
        <span
          class="sr-only"
        >
          Close
        </span>
        <svg
          class="h-5 w-5"
          fill="currentColor"
          viewBox="0 0 20 20"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="bg-white rounded-xl shadow-sm border border-gray-200 "
  >
    <div
      class="px-6 py-4 border-b border-gray-200 "
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="text-lg font-semibold text-gray-900"
        >
          Active Integrations
        </h2>
        <div
          class="flex gap-2"
        >
          <button
            class="px-3 py-1.5 text-sm font-medium rounded-lg transition-colors bg-blue-100 text-blue-700"
          >
            All
          </button>
          <button
            class="px-3 py-1.5 text-sm font-medium rounded-lg transition-colors text-gray-600 hover:bg-gray-100"
          >
            Active
          </button>
          <button
            class="px-3 py-1.5 text-sm font-medium rounded-lg transition-colors text-gray-600 hover:bg-gray-100"
          >
            Inactive
          </button>
        </div>
      </div>
    </div>
    <div
      class="px-6 py-4 p-0"
    >
      <div
        class="text-center py-12"
      >
        <div
          class="flex justify-center mb-4"
        >
          <svg
            class="h-12 w-12 text-gray-400"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 4a2 2 0 114 0v1a1 1 0 001 1h3a1 1 0 011 1v3a1 1 0 01-1 1h-1a2 2 0 100 4h1a1 1 0 011 1v3a1 1 0 01-1 1h-3a1 1 0 01-1-1v-1a2 2 0 10-4 0v1a1 1 0 01-1 1H7a1 1 0 01-1-1v-3a1 1 0 00-1-1H4a2 2 0 110-4h1a1 1 0 001-1V7a1 1 0 011-1h3a1 1 0 001-1V4z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
        <h3
          class="text-lg font-medium text-gray-900 mb-2"
        >
          No integrations found
        </h3>
        <p
          class="text-gray-500 mb-6 max-w-md mx-auto"
        >
          Get started by adding your first exchange rate API integration
        </p>
        <button
          class="inline-flex items-center justify-center font-medium rounded-lg transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:cursor-not-allowed bg-blue-600 text-white hover:bg-blue-700 focus:ring-blue-500 px-4 py-2 text-sm "
          type="button"
        >
          Add Your First Integration
        </button>
      </div>
    </div>
  </div>
</div>
`;
